File: tools/cp2kController/cp2k_shell.c
/**
 *  \file cp2k_shell.c
 *
 *  Functions to control an external cp2k_shell task through std_in/std_out.
 *
 *  The functions returns an int as error value.
 *  
 * \note this file is available also with a BSD style license
 *
 * \copyright (C) 2008  CP2K developers group
 * \author Fawzi Mohamed
 */

#include "cp2k_shell.h"
#include <unistd.h>
#include <stdlib.h>

/**
 * \brief executes (execvp) task with arguments taskArgs in the directory taskDir as subtask.
 *
 * returns file handles std in and std out of the task in \param st
 */
void fm_spawnSubtask(const char *task,const char *taskDir, SubT *st)
{
        int     fdIn[2],fdOut[2];
        pid_t   childpid;

        pipe(fdIn);
        pipe(fdOut);
        
        if((childpid = fork()) == -1)
        {
                perror("fork");
                exit(1);
        }

        if(childpid == 0)
        {
                /* Child process */
				close(fdOut[0]);
                close(fdIn[1]);
				chdir(taskDir);
                dup2(fdIn[0],STDIN_FILENO);
                dup2(fdOut[1],STDOUT_FILENO);
                execlp("sh","sh","-c",task,NULL);
        }
        else
        {
                /* Parent process */
                close(fdIn[0]);
                close(fdOut[1]);
				st->inF=fdopen(fdIn[1],"w");
				st->outF=fdopen(fdOut[0],"r");
				st->isReady=0;
        }
}

/**
 * \brief reads the readiness signal of the task in not already done
 * (asynchronous tasks do not read readiness signal before returning)
 */
int fm_getready_if_needed(SubT *st){
	if (! st->isReady) {
		int err=0;
		err=fm_getready(st);
		if (err) return 200;
	}
	return 0;
}

/**
 * \brief reads the readiness signal of the task
 */
int fm_getready(SubT *st){
	int c,err,writeout=0;
	err=0;
	while ((c=getc(st->outF))!=EOF){
		if (c=='*') break;
		if (c!=' ' && c!='\n' && c!='\t') writeout=1;
		if (writeout) putchar(c); // write out skipped chars
	}
	if (c==EOF) err=-1;
	c=-1;
	if (!err) err=fscanf(st->outF," READY%n",&c)<0;
	if (!err) err=c<0;
	st->isReady= !err;
	return err;
}

/**
 * \brief loads an input file and return the env_id of it
 */
int fm_load(SubT *st,const char *inputF,int *env_id){
	int err;
	err=fm_getready_if_needed(st);
	if (!err) err=fprintf(st->inF,"load %s\n",inputF)<1;
	if (!err) err=fflush(st->inF);
	if (!err) err=fscanf(st->outF,"%d",env_id)!=1;
	if (!err) err=*env_id<1;
	if (!err) err=fm_getready(st);
	return err;
}

/**
 * \brief Begins loading an input file.
 *
 * Asynchronous, does wait or for completion.
 */
int fm_bg_load(SubT *st,const char *inputF){
	int err;
	err=fm_getready_if_needed(st);
	if (!err) err=fprintf(st->inF,"bg_load %s\n",inputF)<1;
	if (!err) err=fflush(st->inF);
	st->isReady=0;
	return err;
}

/**
 * \brief returns the env_id of the last file loaded
 */
int fm_last_env_id(SubT *st,int *env_id){
	int err;
	err=fm_getready_if_needed(st);
	if (!err) err=fprintf(st->inF,"last_env_id\n")<1;
	if (!err) err=fflush(st->inF);
	if (!err) err=fscanf(st->outF,"%d",env_id)!=1;
	if (!err) err=*env_id<1;
	if (!err) err=fm_getready(st);
	return err;
}

/**
 * \brief returns number of atoms in the given env_id
 */
int fm_natom(SubT *st,int env_id,int *natom){
	int err;
	err=fm_getready_if_needed(st);
	if (!err) {
		if (env_id>0) {
			err=fprintf(st->inF,"natom %d\n",env_id)<1;
		} else {
			err=fprintf(st->inF,"natom\n")<1;
		}
	}
	if (!err) err=fflush(st->inF);
	if (!err) err=fscanf(st->outF,"%d",natom)!=1;
	if (!err) err=fm_getready(st);
	return err;
}
	
/**
 * \brief changes the position of the atoms of the given env_id
 */
int fm_setpos(SubT *st,int env_id,int npos,const double pos[]){
	int err,i;
	err=fm_getready_if_needed(st);
	if (!err) {
		if (env_id>0) {
			err=fprintf(st->inF,"setpos %d\n",env_id)<1;
		} else {
			err=fprintf(st->inF,"setpos\n")<1;
		}
	}
	if (!err) err=fprintf(st->inF,"%d\n",npos)<1;
	for (i=0;i<npos;i++){
		if (!err) err=fprintf(st->inF," %22.13g",pos[i])<1;
	}
	if (!err) err=fprintf(st->inF,"\n* END\n")<1;
	if (!err) err=fflush(st->inF);
	if (!err) err=fm_getready(st);
	return err;
}

/**
 * \brief returns the position of the atoms of the given env_id
 */
int fm_getpos(SubT *st,int env_id,int npos,double pos[]){
	int err, npos2,i,c;
	err=fm_getready_if_needed(st);
	if (!err){
		if (env_id>0) {
			err=fprintf(st->inF,"getpos %d\n",env_id)<1;
		} else {
			err=fprintf(st->inF,"getpos\n")<1;
		}
	}
	if (!err) err=fflush(st->inF);
	if (!err) err=fscanf(st->outF,"%d",&npos2)!=1;
	if (!err) err=npos!=npos2;
	for (i=0;i<npos;i++){
		if (!err) err=fscanf(st->outF,"%lg",&pos[i])!=1;
	}
	while ((c=getc(st->outF))!=EOF){
		if (c=='*') break;
	}
	c=-1;
	if (!err) err=fscanf(st->outF," END%n",&c);
	if (!err) err=c<0;
	if (!err) err=fm_getready(st);
	return err;
}

/**
 * \brief begins the computation of e and f.
 * 
 * Asynchronous, does not wait for completion
 */
int fm_eval_ef(SubT *st,int env_id){
	int err;
	err=fm_getready_if_needed(st);
	if (!err) {
		if (env_id>0) {
			err=fprintf(st->inF,"eval_ef %d\n",env_id)<1;
		} else {
			err=fprintf(st->inF,"eval_ef\n")<1;
		}
	}
	if (!err) err=fflush(st->inF);
	st->isReady=0;
	return err;
}

/**
 * \brief gets the force of the last force computation.
 * \warning if you calculated just the energy, the content of the array are undefined.
 */
int fm_get_f(SubT *st,int env_id,int nforce,double force[]){
	int err, nforce2, i, c;
	err=fm_getready_if_needed(st);
	if (!err) {
		if (env_id>0) {
			err=fprintf(st->inF,"get_f %d\n",env_id)<1;
		} else {
			err=fprintf(st->inF,"get_f\n")<1;
		}
	}
	if (!err) err=fflush(st->inF);
	if (!err) err=fscanf(st->outF,"%d",&nforce2)!=1;
	if (!err) err=nforce!=nforce2;
	for (i=0;i<nforce;i++){
		if (!err) err=fscanf(st->outF,"%lg",&force[i])!=1;
	}
	while ((c=getc(st->outF))!=EOF){
		if (c=='*') break;
	}
	c=-1;
	if (!err) err=fscanf(st->outF," END%n",&c);
	if (!err) err=c<1;
	if (!err) err=fm_getready(st);
	return err;
}


/**
 * \brief Calculates the energy and forces and returns them
 */
int fm_calc_ef(SubT *st,int env_id,double *energy,int nforce,double force[]){
	int err,nforce2,i,c;
	err=fm_getready_if_needed(st);
	if (!err) {
		if (env_id>0) {
			err=fprintf(st->inF,"calc_ef %d\n",env_id)<1;
		} else {
			err=fprintf(st->inF,"calc_ef\n")<1;
		}
	}
	if (!err) err=fflush(st->inF);
	if (!err) err=fscanf(st->outF,"%lg",energy)!=1;
	if (!err) err=fscanf(st->outF,"%d",&nforce2)!=1;
	if (!err) err=nforce!=nforce2;
	for (i=0;i<nforce;i++){
		if (!err) err=fscanf(st->outF,"%lg",&force[i])!=1;
	}
	while ((c=getc(st->outF))!=EOF){
		if (c=='*') break;
	}
	c=-1;
	if (!err) err=fscanf(st->outF," END%n",&c);
	if (!err) err=c<1;
	if (!err) err=fm_getready(st);
	return err;
}

/**
 * \brief begins the computation of e.
 * 
 * Asynchronous, does not wait for completion
 * \note the force of the environment is undefined after this call
 */
int fm_eval_e(SubT *st,int env_id){
	int err;
	err=fm_getready_if_needed(st);
	if (!err) {
		if (env_id>0) {
			err=fprintf(st->inF,"eval_e %d\n",env_id)<1;
		} else {
			err=fprintf(st->inF,"eval_e\n")<1;
		}
	}
	if (!err) err=fflush(st->inF);
	st->isReady=0;
	return err;
}

/**
 * \brief Calculates the energy and returns it
 *
 * the force of the environment is undefined after this call
 */
int fm_calc_e(SubT *st,int env_id,double *energy){
	int err;
	if (env_id>0) {
		err=fprintf(st->inF,"calc_e %d\n",env_id)<1;
	} else {
		err=fprintf(st->inF,"calc_e\n")<1;
	}
	if (!err) err=fflush(st->inF);
	if (!err) err=fscanf(st->outF,"%lg",energy)!=1;
	if (!err) err=fm_getready(st);
	return err;
}

/**
 * \brief returns the energy of the last computation
 */
int fm_get_e(SubT *st,int env_id,double *energy){
	int err;
	err=fm_getready_if_needed(st);
	if (!err) {
		if (env_id>0) {
			err=fprintf(st->inF,"get_e %d\n",env_id)<1;
		} else {
			err=fprintf(st->inF,"get_e\n")<1;
		}
	}
	if (!err) err=fflush(st->inF);
	if (!err) err=fscanf(st->outF,"%lg",energy)!=1;
	if (!err) err=fm_getready(st);
	return err;
}

/**
 * \brief returns some informations about the underlying cp2k task
 *
 * \note assumes that no line begins with '*'
 */
int fm_get_info(SubT *st,int sizeBuf,char *buf,int *totalSize){
	int err,nread=0,atStart=1,c,totc=1;
	err=fm_getready_if_needed(st);
	if (!err) err=fprintf(st->inF,"info\n")<1;
	if (!err) err=fflush(st->inF);
	while ((c=getc(st->outF))!=EOF){
		if (atStart && c=='*') break;
		if (c!=' ' && c!='\t') atStart=0;
		if (c=='\n' || c=='\r') atStart=1;
		if (nread<sizeBuf) {
			buf[nread]=(char)c;
			nread++;
		}
		totc++;
	}
	if (nread<sizeBuf) nread++;
	if (nread<sizeBuf) buf[nread]='\0';
	if (c==EOF) err=-1;
	c=-1;
	if (!err) err=fscanf(st->outF," READY%n",&c)<0;
	if (!err) err=c<0;
	st->isReady= !err;
	*totalSize=totc;
	return err;
}

File: tools/cp2kController/cp2k_shell.py
import subprocess,os,os.path,re
from numpy import *

class Cp2kEnv:
    def __init__(self,cpshell,env_id):
        self.cpshell=cpshell
        self.env_id=env_id
    def eval_e(self):
        'evaluates the energy in background'
        nerr=self.cpshell.nErrors
        self.cpshell.exec_cmd('eval_e %d\n'%(self.env_id),async=1)
        self.cpshell.raiseLastError(nerr)
    def eval_ef(self):
        'evaluates the energy and forces in background'
        nerr=self.cpshell.nErrors
        self.cpshell.exec_cmd('eval_ef %d\n'%(self.env_id),async=1)
        self.cpshell.raiseLastError(nerr)
    def get_pos(self):
        'returns the actual position of the atoms'
        nerr=self.cpshell.nErrors
        lines=self.cpshell.exec_cmd('get_pos %d\n'%(self.env_id))
        self.cpshell.raiseLastError(nerr)
        ndim=int(lines[0])
        pos=array(map(float,("".join(lines[1:])).split()[:ndim]))
        return reshape(pos,(ndim/3,3))
    def get_f(self):
        'returns the force of the atoms'
        nerr=self.cpshell.nErrors
        lines=self.cpshell.exec_cmd('get_f %d\n'%(self.env_id))
        self.cpshell.raiseLastError(nerr)
        ndim=int(lines[0])
        f=array(map(float,("".join(lines[1:])).split()[:ndim]))
        return reshape(f,(ndim/3,3))
    def get_natom(self):
        '''returns the number of atoms in the environment'''
        nerr=self.cpshell.nErrors
        lines=self.cpshell.exec_cmd('natom %d\n'%(self.env_id))
        self.cpshell.raiseLastError(nerr)
        return int(lines[0])
    def get_e(self):
        '''returns the energy, of the last configuaration evaluated,
         0.0 if none was evaluated'''
        nerr=self.cpshell.nErrors
        lines=self.cpshell.exec_cmd('get_e %d\n'%(self.env_id))
        self.cpshell.raiseLastError(nerr)
        return float(lines[0])
    def set_pos(self,pos):
        """changes the position of the environement
        returns the maximum change in coordinates"""
        nerr=self.cpshell.nErrors
        fin=self.cpshell.st.stdin
        fin.write('set_pos %d\n'%(self.env_id))
        p=ravel(pos)
        fin.write("%d\n"%(p.size))
        for pp in p:
            fin.write(' %18.13f'%(pp))
        fin.write("\n* END\n")
        fin.flush()
        lines=self.cpshell.getReady()
        self.cpshell.raiseLastError(nerr)
        return float(lines[0])
    def destroy(self):
        "destroys this environment"
        nerr=self.cpshell.nErrors
        lines=self.cpshell.exec_cmd('destroy %d\n'%(self.env_id))
        self.cpshell.envs[self.env_id]
        self.env_id=-1
        self.cpshell.raiseLastError(nerr)

class Cp2kShell:
    def __init__(self,task,taskDir):
    	"starts a cp2k task"
    	myDir=os.getcwd()
    	os.chdir(taskDir)
    	self.st=subprocess.Popen(task,shell=True,
    		stdin=subprocess.PIPE,
    		stdout=subprocess.PIPE,
    		stderr=subprocess.STDOUT,
    		close_fds=1, env=os.environ)
    	self.isReady=0
    	self.envs={}
    	self.nErrors=0
    	self.lastErrors=[]
    	os.chdir(myDir)
    
    def raiseLastError(self,nErrors=0):
        if nErrors!=self.nErrors:
            raise Exception("".join(self.lastErrors))
    
    def addErrs(self,errs):
        'adds the given errors to the task errors'
        if errs:
            self.nErrors+=1
            self.lastErrors=errs
    
    def clearErrs(self):
        self.nErrors=0
        self.lastErrors=[]
    
    def getReady(self):
        "reads the readiness message of the subtask"
        readyRe=re.compile(r"^ *\* *(?:ready).*",re.IGNORECASE)
        errorRe=re.compile(r"^ *\* *error.*",re.IGNORECASE)
        lines=[]
        errors=[]
        while 1:
            line=self.st.stdout.readline()
            if not line: break
            if errorRe.match(line):
                errors.append(line)
            if readyRe.match(line):
                self.isReady=1
                self.addErrs(errors)
                return lines
            lines.append(line)
        self.addErrs(errors)        
        raise Exception('cp2k_shell did not get ready')
    
    def exec_cmd(self,cmdStr,async=0):
        "executes a command and (if async) returns the output"
        if not self.isReady:
            lines=self.getReady()
        self.st.stdin.write(cmdStr)
        if cmdStr[-1]!='\n':
            self.st.stdin.write('\n')            
        self.st.stdin.flush()
        if async:
            self.isReady=0
        else:
            return self.getReady()
    
    def newEnv(self,inputFile,runDir=""):
        """creates a new environment
        changes the current working directory to it"""
        nerr=self.nErrors
        if runDir and runDir!=".":
            self.exec_cmd('cd '+runDir)
        if self.nErrors==nerr:
            lines=self.exec_cmd('load '+inputFile)
            if self.nErrors==nerr:
                env_id=int(lines[0])
                self.envs[env_id]=Cp2kEnv(self,env_id)
                return self.envs[env_id]
        return None
    
    def newBgEnv(self,inputFile,runDir=""):
        "creates a new environment in background, use lastEnv to get it"
        nerr=self.nErrors
        if runDir and runDir!=".":
            self.exec_cmd('cd '+runDir)
        if self.nErrors==nerr:
            self.exec_cmd('bg_load '+inputFile,async=1)
    
    def lastEnv(self):
        "returns the last environment created"
        lines=self.exec_cmd('last_env_id')
        env_id=int(lines[0])
        if env_id<=0:
            return None
        if not self.envs.has_key(env_id):
            self.envs[env_id]=Cp2kEnv(self,env_id)
        return self.envs[env_id]
    
    def chdir(self,newDir):
        "changes the working directory"
        self.exec_cmd('cd '+newDir)
    
    def getcwd(self):
        "returns the working directory"
        return self.exec_cmd('pwd')[0][:-1]
        
        